### PR TITLE
PERF: maybe_convert_numeric speedup

### PR DESCRIPTION
**Pull request by chris-b1**
xref #16043
 
Probably more to be done, but picks up some easy performance on the int path.  For some reason I'm having trouble running asv, here are some timings.

### setup
```
ints = np.array(list(range(1000000)), dtype=object)
floats = np.array([float(x) for x in range(1000000)], dtype=object)
ints_with_float = ints.copy()
ints_with_float[-1] = 1.2
```
### Master
```
In [2]: %timeit pd.to_numeric(floats)
10 loops, best of 3: 132 ms per loop

In [3]: %timeit pd.to_numeric(ints)
1 loop, best of 3: 437 ms per loop

In [4]: %timeit pd.to_numeric(ints_with_float)
1 loop, best of 3: 444 ms per loop
```

### PR
```
In [5]: %timeit pd.to_numeric(floats)
10 loops, best of 3: 88.9 ms per loop

In [6]: %timeit pd.to_numeric(ints)
10 loops, best of 3: 57.6 ms per loop

In [7]: %timeit pd.to_numeric(ints_with_float)
1 loop, best of 3: 392 ms per loop
```

### CODE REVIEW

File: pandas/_libs/src/inference.pyx [hunk 1]
```diff
@@ -970,13 +974,12 @@ def maybe_convert_numeric(ndarray[object] values, set na_values,
...
 
-            as_int = int(val)
-            seen.saw_int(as_int)
+            seen.saw_int(val)
 
-            if as_int >= 0:
-                uints[i] = as_int
-            if as_int <= iINT64_MAX:
-                ints[i] = as_int
+            if val >= 0:
```
Comment (jreback) [inline pandas/_libs/src/inference.pyx]: you might be able to 

```
cdef ival <int> val
if ival >= 0:
     ...
if ival <= iINT64_max:
    ..
```

**Comment by jreback** [comment]
the other thing is it *might* be worth it to actually try to cast once you have found an actual int (or a uint), 

```
In [13]: (ints.astype('int64', casting='unsafe') == ints).all()
Out[13]: True

In [14]: %timeit (ints.astype('int64', casting='unsafe') == ints).all()
55.5 ms ± 1.15 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

At this point it *would* raise if you encounter a float so this is ok (you can do it in a try/except).

not sure if this is totally feasible though.

File: pandas/_libs/src/inference.pyx (hunk 1 above)
Comment (chris-b1) [inline pandas/_libs/src/inference.pyx]: I can't do exactly this because val could be larger than fits into a `c` int

**Comment by chris-b1** [comment]
OK, I've added some fastpath logic that will try and convert the whole array at once - have to see if anything breaks!

**Comment by TomAugspurger** [comment]
This should go in 0.20, correct?

**Comment by chris-b1** [comment]
I need to work through a couple more tests, but assuming I can get this to
work in a non-breaking way, yes I'd think this would be good for 0.20.
`to_numeric` performance with ints has regressed quite a bit since 0.19.1
(although is more functional now, to be fair)

**Comment by jreback** [comment]
nice! thanks @chris-b1

**Comment by jorisvandenbossche** [comment]
Yeah, I put the 0.21 milestone on it as it was about performance, but forgot it was for a perf regressiono in 0.20dev. So certainly fine. Thanks Chris!

**Pull request merged**